### PR TITLE
PROJECTION expression

### PR DESCRIPTION
**Pull request by n9iels**
This PR adds support for the `PROJECTION`, an additional expression that can be used together with the `CREATE` expression.

The  expression consists out of the properties `context` and `fact`. The `context` property is an array to allow the usage of nested `CREATE` expressions.
```json
{
    "expression": "PROJECTION",
    "context": [
        "[aanvraag]"
    ],
    "fact": "[bedrag]"
}
```

Model example:
```json
{
    "acts": [
        {
            "act": "<<subsidie aanvragen>>",
            "actor": "[burger]",
            "recipient": "[ambtenaar]",
            "object": "[verzoek]",
            "preconditions": {
                "expression": "AND",
                "operands": [
                    "[bedrag]"
                ]
            },
            "create": [
                "[aanvraag]"
            ]
        },
        {
            "act": "<<subsidie aanvraag toekennen>>",
            "actor": "[ambtenaar]",
            "object": "[aanvraag]",
            "preconditions": {
                "expression": "LESS_THEN",
                "operands": [
                    {
                        "expression": "PROJECTION",
                        "context": [
                            "[aanvraag]"
                        ],
                        "fact": "[bedrag]"
                    },
                    {
                        "expression": "LITERAL",
                        "operand": 500
                    }
                ]
            },
            "recipient": "[burger]"
        }
    ],
    "facts": [
        {
            "fact": "[aanvraag]",
            "function": {
                "expression": "CREATE",
                "operands": [
                    "[bedrag]"
                ]
            }
        }
    ],
    "duties": []
}
```

### CODE REVIEW

File: test/index.spec.js [hunk 1]
```diff
@@ -1941,4 +1888,450 @@ describe('discipl-law-reg', () => {
...
+    it('should project a series of CREATE facts', async () => {
+      const model = {
+        'acts': [
+          {
+            'act': '<<persoonlijk gegevens invullen>>',
+            'actor': '[burger]',
+            'recipient': '[ambtenaar]',
+            'object': '[verzoek]',
+            'preconditions': {
+              'expression': 'AND',
```
Comment (pimotte) [inline test/index.spec.js:2172]: Any particular reason this has the "AND" operator wrapping the operand?

**Comment by n9iels** [comment]
@pimotte do these new expressions need to be implemented in https://github.com/discipl/flinteditor-vscode too?

File: test/flint-example-lerarenbeurs.json [hunk 2]
```diff
@@ -959,7 +959,10 @@
     },
     {
       "fact": "[besluit]",
-      "function": { "expression": "CREATE" },
+      "function": {
+        "expression": "CREATE",
+        "operands": []
```
Comment (pimotte) [inline test/flint-example-lerarenbeurs.json:964]: Do you think we should make operands optional? I think the trade-off here is that we don't have this optionality in other places, but it avoids introducing a breaking change.

**Comment by pimotte** [comment]
> @pimotte do these new expressions need to be implemented in https://github.com/discipl/flinteditor-vscode too?

I'm tempted to leave it and instead work with @blackstardlb to get it supported in flinteditor-mps instead.

File: test/index.spec.js (hunk 1 above)
Comment (n9iels) [reply test/index.spec.js:2172]: No :) Usage of `AND` is indeed redundant in this case

File: test/flint-example-lerarenbeurs.json (hunk 2 above)
Comment (n9iels) [reply test/flint-example-lerarenbeurs.json]: I agree. Also, the expression it useful without any operands. So it makes sense to me

**Comment by n9iels** [comment]
> Document the new operations

I have added some documentation to https://github.com/discipl/discipl-law-reg/pull/116 and made it ready for review.

> Test for OR in CREATE expression

I have added a test where a `OR` expression is used withing `CREATE` expressions.

**Review by pimotte — APPROVED** [review]
Reviewed this before, content is good:D

Style/structure improvements across the board I'm happy with too, well done! Props to both of you, @n9iels and @blackstardlb

**Pull request merged**